## You're here! 
Welcome to your first competition in the [ITI's AI Pro training program](https://ai.iti.gov.eg/epita/ai-engineer/)! We hope you enjoy and learn as much as we did prepairing this competition.


## Introduction

In the competition, it's required to predict the `Severity` of a car crash given info about the crash, e.g., location.

This is the getting started notebook. Things are kept simple so that it's easier to understand the steps and modify it.

Feel free to `Fork` this notebook and share it with your modifications **OR** use it to create your submissions.

### Prerequisites
You should know how to use python and a little bit of Machine Learning. You can apply the techniques you learned in the training program and submit the new solutions! 

### Checklist
You can participate in this competition the way you perefer. However, I recommend following these steps if this is your first time joining a competition on Kaggle.

* Fork this notebook and run the cells in order.
* Submit this solution.
* Make changes to the data processing step as you see fit.
* Submit the new solutions.

*You can submit up to 5 submissions per day. You can select only one of the submission you make to be considered in the final ranking.*


Don't hesitate to leave a comment or contact me if you have any question!

## Import the libraries

We'll use `pandas` to load and manipulate the data. Other libraries will be imported in the relevant sections.

In [1]:
import pandas as pd
import os

## Exploratory Data Analysis
In this step, one should load the data and analyze it. However, I'll load the data and do minimal analysis. You are encouraged to do thorough analysis!

Let's load the data using `pandas` and have a look at the generated `DataFrame`.

In [2]:
dataset_path = '/kaggle/input/car-crashes-severity-prediction/'

df = pd.read_csv(os.path.join(dataset_path, 'train.csv'))

print("The shape of the dataset is {}.\n\n".format(df.shape))

df.head()

The shape of the dataset is (6407, 16).




ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0   0  37.762150 -122.405660  False         0.044     False     False   
1   1  37.719157 -122.448254  False         0.000     False     False   
2   2  37.808498 -122.366852  False         0.000     False     False   
3   3  37.785930 -122.391080  False         0.009     False     False   
4   4  37.719141 -122.448457  False         0.000     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  Severity  \
0     False    False    False       False  False     True    R         2   
1     False    False    False       False  False    False    R         2   
2     False    False    False       False   True    False    R         3   
3      True    False    False       False  False    False    R         1   
4     False    False    False       False  False    False    R         2   

             timestamp  
0  2016-03-25 15:13:02  
1  2020-05-05 19:23:00  
2  2016-09-16 19:57:16  
3  2020-03-29 19:48:43  
4  2019-10-09 08:47:00

In [3]:
weather_df = pd.read_csv(os.path.join(dataset_path, 'weather-sfcsv.csv'))
weather_df["date"] = pd.to_datetime(weather_df[["Year","Day","Month","Hour"]]).dt.date
weather_df["time"] = pd.to_datetime(weather_df[["Year","Day","Month","Hour"]]).dt.hour
# weather_df["timestamp"] = pd.to_datetime(weather_df[["Year","Day","Month","Hour"]])
weather_df = weather_df.drop(columns = ['Year','Day','Month','Hour'])
weather_df = weather_df.sort_values(by=['date'])
# weather_df['timestamp']=pd.Series(weather_df['timestamp']).dt.round("6H")

weather_df=weather_df.drop(columns=['Precipitation(in)','Selected'])
weather_df=weather_df.drop(columns=['Weather_Condition','Wind_Chill(F)'])
weather_df.dropna(subset = ['Wind_Speed(mph)','Visibility(mi)','Humidity(%)','Temperature(F)'],inplace=True)
# weather_df=weather_df.groupby(by=["timestamp"]).mean()
weather_df.drop_duplicates(subset = ["date","time"], keep='first', inplace=True)
weather_df.head()

Temperature(F)  Humidity(%)  Wind_Speed(mph)  Visibility(mi)  \
1053            61.0         62.0             20.7            10.0   
3329            59.0         67.0             12.7            10.0   
3813            55.0         80.0              6.9            10.0   
401             55.9         77.0              3.5            10.0   
2890            57.0         83.0             20.7            10.0   

            date  time  
1053  2016-03-23    16  
3329  2016-03-23    20  
3813  2016-03-24     8  
401   2016-03-24     9  
2890  2016-03-24    19

In [4]:
import xml.etree.ElementTree as ET
path=os.path.join(dataset_path, 'holidays.xml')
xml_data = open(path, 'r').read()  # Read file
root = ET.XML(xml_data)
data = []
cols = []
for i, child in enumerate(root):
    data.append([subchild.text for subchild in child])
    cols.append(child.tag)

dfHol = pd.DataFrame(data).T  # Write in DF and transpose it

#dfHol.columns = cols  # Update column names


dfHol=dfHol.T
dfHol.columns=['date','holiday']

dfHol['date']=pd.to_datetime(dfHol['date']).dt.date

In [5]:

def dataPrep(df):
    df['timestamp']=pd.to_datetime(df['timestamp'])
    df['date']=df['timestamp'].dt.date
    df['time']=df['timestamp'].dt.hour
    dfAll=pd.merge(df,dfHol,how='left' ,on='date')
    dfAll=dfAll.drop(columns=['Bump','Roundabout','No_Exit','Give_Way'])
    dfAll['holiday'].fillna(0,inplace=True)
    dfAll['holiday']=dfAll['holiday'].apply(lambda x: 1 if x != 0 else x)
    dfAll.replace({False: 0, True: 1}, inplace=True)
    dfAll.replace({'R': 0, 'L': 1}, inplace=True)
    dfAll.drop_duplicates(subset = ["ID"], keep='first', inplace=True)
    dfAll=pd.merge(dfAll,weather_df,how='left' ,on=['date','time'])
    #dfAll.dropna(subset = ['Wind_Speed(mph)','Visibility(mi)','Humidity(%)','Temperature(F)'],inplace=True)
    dfAll.drop(columns='timestamp',inplace=True)
    dfAll['date']=pd.DatetimeIndex(dfAll['date']).year
    dfAll.fillna((dfAll.mean()), inplace=True)
    return dfAll
    
df = dataPrep(df)





   









We've got 6407 examples in the dataset with 14 featues, 1 ID, and the `Severity` of the crash.

By looking at the features and a sample from the data, the features look of numerical and catogerical types. What about some descriptive statistics?

The output shows desciptive statistics for the numerical features, `Lat`, `Lng`, `Distance(mi)`, and `Severity`. I'll use the numerical features to demonstrate how to train the model and make submissions. **However you shouldn't use the numerical features only to make the final submission if you want to make it to the top of the leaderboard.**

## Data Splitting

Now it's time to split the dataset for the training step. Typically the dataset is split into 3 subsets, namely, the training, validation and test sets. In our case, the test set is already predefined. So we'll split the "training" set into training and validation sets with 0.8:0.2 ratio. 

*Note: a good way to generate reproducible results is to set the seed to the algorithms that depends on randomization. This is done with the argument `random_state` in the following command* 

In [6]:
from sklearn.model_selection import train_test_split

train_df, val_df = train_test_split(df, test_size=0.2, random_state=42) # Try adding `stratify` here

X_train = train_df.drop(columns=['ID', 'Severity'])
y_train = train_df['Severity']

X_val = val_df.drop(columns=['ID', 'Severity'])
y_val = val_df['Severity']


As pointed out eariler, I'll use the numerical features to train the classifier. **However, you shouldn't use the numerical features only to make the final submission if you want to make it to the top of the leaderboard.** 

## Model Training

Let's train a model with the data! We'll train a Random Forest Classifier to demonstrate the process of making submissions. 

In [7]:
from sklearn.ensemble import RandomForestClassifier

# Create an instance of the classifier
classifier = RandomForestClassifier(max_depth=2, random_state=0)

# Train the classifier
classifier = classifier.fit(X_train, y_train)

Now let's test our classifier on the validation dataset and see the accuracy.

In [8]:
print("The accuracy of the classifier on the validation set is ", (classifier.score(X_val, y_val)))

The accuracy of the classifier on the validation set is  0.7457098283931357


Well. That's a good start, right? A classifier that predicts all examples' `Severity` as 2 will get around 0.63. You should get better score as you add more features and do better data preprocessing.

## Submission File Generation

We have built a model and we'd like to submit our predictions on the test set! In order to do that, we'll load the test set, predict the class and save the submission file. 

First, we'll load the data.

In [9]:
test_df = pd.read_csv(os.path.join(dataset_path, 'test.csv'))
test_df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0  6407  37.786060 -122.390900  False         0.039     False     False   
1  6408  37.769609 -122.415057  False         0.202     False     False   
2  6409  37.807495 -122.476021  False         0.000     False     False   
3  6410  37.761818 -122.405869  False         0.000     False     False   
4  6411  37.732350 -122.414100  False         0.670     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  \
0      True    False    False       False  False    False    R   
1     False    False    False       False  False    False    R   
2     False    False    False       False  False    False    R   
3      True    False    False       False  False    False    R   
4     False    False    False       False  False    False    R   

             timestamp  
0  2016-04-04 19:20:31  
1  2020-10-28 11:51:00  
2  2019-09-09 07:36:45  
3  2019-08-06 15:46:25  
4  2018-10-17 09:54:58

Note that the test set has the same features and doesn't have the `Severity` column.
At this stage one must **NOT** forget to apply the same processing done on the training set on the features of the test set.

Now we'll add `Severity` column to the test `DataFrame` and add the values of the predicted class to it.

**I'll select the numerical features here as I did in the training set. DO NOT forget to change this step as you change the preprocessing of the training data.**

In [10]:
test_df=dataPrep(test_df)

X_test = test_df.drop(columns=['ID'])
X_test
#

# You should update/remove the next line once you change the features used for training
#X_test = X_test[['Lat', 'Lng', 'Distance(mi)']]
#//
y_test_predicted = classifier.predict(X_test)

test_df['Severity'] = y_test_predicted

test_df.head()

ID        Lat         Lng  Distance(mi)  Crossing  Junction  Railway  \
0  6407  37.786060 -122.390900         0.039         0         1        0   
1  6408  37.769609 -122.415057         0.202         0         0        0   
2  6409  37.807495 -122.476021         0.000         0         0        0   
3  6410  37.761818 -122.405869         0.000         0         1        0   
4  6411  37.732350 -122.414100         0.670         0         0        0   

   Stop  Amenity  Side  date  time  holiday  Temperature(F)  Humidity(%)  \
0     0        0     0  2016    19        0            63.0         60.0   
1     0        0     0  2020    11        0            65.0         56.0   
2     0        0     0  2019     7        0            57.0         87.0   
3     0        0     0  2019    15        0            72.0         59.0   
4     0        0     0  2018     9        0            57.0         77.0   

   Wind_Speed(mph)  Visibility(mi)  Severity  
0             10.4            10.0         2  
1              5.0             9.0         2  
2              3.0             6.0         2  
3             17.0            10.0         2  
4              5.8            10.0         2

Now we're ready to generate the submission file. The submission file needs the columns `ID` and `Severity` only.

In [11]:
test_df[['ID', 'Severity']].to_csv('/kaggle/working/submission.csv', index=False)

The remaining steps is to submit the generated file and are as follows. 

1. Press `Save Version` on the upper right corner of this notebook.
2. Write a `Version Name` of your choice and choose `Save & Run All (Commit)` then click `Save`.
3. Wait for the saved notebook to finish running the go to the saved notebook.
4. Scroll down until you see the output files then select the `submission.csv` file and click `Submit`.

Now your submission will be evaluated and your score will be updated on the leaderboard! CONGRATULATIONS!!

## Conclusion

In this notebook, we have demonstrated the essential steps that one should do in order to get "slightly" familiar with the data and the submission process. We chose not to go into details in each step to keep the welcoming notebook simple and make a room for improvement.

You're encourged to `Fork` the notebook, edit it, add your insights and use it to create your submission.